ports/stm32/eth.c

```diff
@@ -211,13 +201,30 @@ STATIC int eth_mac_init(eth_t *self) {
     mp_hal_pin_config_alt_static(MICROPY_HW_ETH_RMII_TXD0, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, STATIC_AF_ETH_RMII_TXD0);
     mp_hal_pin_config_alt_static(MICROPY_HW_ETH_RMII_TXD1, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, STATIC_AF_ETH_RMII_TXD1);
 
+    // Enable peripheral clock
     #if defined(STM32H7)
     __HAL_RCC_ETH1MAC_CLK_ENABLE();
     __HAL_RCC_ETH1TX_CLK_ENABLE();
     __HAL_RCC_ETH1RX_CLK_ENABLE();
-    __HAL_RCC_ETH1MAC_FORCE_RESET();
     #else
     __HAL_RCC_ETH_CLK_ENABLE();
+    #endif
+}
+
+void eth_set_trace(eth_t *self, uint32_t value) {
+    self->trace_flags = value;
+}
+
+STATIC int eth_mac_init(eth_t *self) {
+    // Configure MPU
+    uint32_t irq_state = mpu_config_start();
+    mpu_config_region(MPU_REGION_ETH, (uint32_t)&eth_dma, MPU_CONFIG_ETH(MPU_REGION_SIZE_16KB));
+    mpu_config_end(irq_state);
+
+    // Reset Eth peripheral.
+    #if defined(STM32H7)
```

Comment:
Won't this need to enable the clock in case it was put in low power mode?  And also disable low-power mode in the BCR (or is the controller reset below)?

Reply:
The reset takes it out of low-power mode, but it does need the clock re-enabled if it was disabled.

Reply:
Updated.

Reply:
Actually that might not have been absolutely necessary... If eth was put in low-power mode from say board_early_init, and then used normally from Python, eth_init() is called from make_new which will re-enable the clock... I can remove it again if you like.

```
STATIC mp_obj_t network_lan_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
    mp_arg_check_num(n_args, n_kw, 0, 0, false);
    const network_lan_obj_t *self = &network_lan_eth0;
    eth_init(self->eth, MP_HAL_MAC_ETH0); <--- Will re-enable the clock if in low-power mode.
```

Reply:
I just tested this again it and reset does Not exit low-power mode, so I added a line in eth_start() to exit low-power mode. If you prefer I could set a flag when in low-power mode instead of calling it unconditionally.

Reply:
It looks good as it is.